### PR TITLE
[RNMobile] Auto-scroll upon block insertion

### LINKED ISSUE #2773: Content doesn't scroll to new block upon inserting one
**Describe the bug**
When inserting a block that doesn't auto-focus on insertion (e.g. Image, Video, etc), the UI doesn't scroll to the new block at the end of the context, when there is enough content already to fill the screen.

**To Reproduce**
1. Open the demo app on Android or iOS
2. Tap on the Inserter ("+") and add a new block that doesn't auto-focus on insertion (e.g. Image, Video, etc)
3. Notice that the UI doesn't scroll to the end of the content where the new block was appended

**Expected behavior**
Inserting a new Paragraph block for example causes the UI to scroll to the end of the content when the new block is created (even when the insertion happens off-screen). The same behavior should apply to all other types of blocks (e.g. Image, Video, etc).

**Screenshots**

| iOS screen recording  | Android screen recording   |
| - | - |

**Smartphone:**
 - Device: Tested on Samsung S10 and iPhone 11
 - OS: Android 10 and iOS 14
 - Version https://github.com/WordPress/gutenberg/commit/40a5502d62534cd64a623938754778501264736c

### PR DESCRIPTION
**Pull request by fluiddot**
**Related PRs:**
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6482
* 🧪 For testing:
    * https://github.com/wordpress-mobile/WordPress-Android/pull/19832
   * https://github.com/wordpress-mobile/WordPress-iOS/pull/22266

## What?

Add a mechanism to automatically scroll to a block after being inserted.

## Why?

* Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/2773.
* Enhances block insertion experience by helping users continue editing content where they just inserted a block.

## How?

This section is divided into subsections related to the three main items tackled in this PR:
1. The conditions to trigger the auto-scroll.
2. The component to check the conditions.
3. Determine the offset for scrolling.

### Conditions to trigger the auto-scroll

* https://github.com/WordPress/gutenberg/pull/57273/commits/bbe30e358e73c7567c1245ce4965061ccab99645

I assumed we needed two conditions for this:
* The block is selected: This is widely checked in different components.
* The block was just inserted: We already have a selector to identify the last blocks inserted ([`wasBlockJustInserted`](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/data/data-core-block-editor.md#wasblockjustinserted)).

Based on the above, the following cases will trigger the auto-scroll:
- Insert a block via the inserter menu.
- Copy and paste a block via the inserter menu.
- Insert a block with the following insertion options:
    - Add To Beginning.
    - Add Block Before.
    - Add Block After.
    - Add To End.

And these won't trigger it:
- Insert a block via the slash inserter.
- Insert blocks by pasting HTML content.
- Copy and paste a block via the block options bottom sheet.
- Duplicate a block.

### The component to check the conditions

* https://github.com/WordPress/gutenberg/pull/57273/commits/bbe30e358e73c7567c1245ce4965061ccab99645

I was hesitant about whether to implement the check in the `BlockList` or the `Block` components. I finally decided to use the `Block` component due to the need to retrieve the layout data for calculating the scroll offset (see more in the below section).

Most of the logic related to this has been extracted in a new hook named `useScrollUponInsertion`.

### Determine the offset for scrolling

The calculation of the offset requires four elements:
- Y-position of the block being inserted.
- Height of the block being inserted.
- Current scroll offset.
- Height of scroll view.

#### Reuse scroll logic from `KeyboardAwareFlatList`
A similar calculation is being performed to scroll when typing on a `RichText` component ([reference](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/keyboard-aware-flat-list/use-scroll-to-text-input.native.js)), so part of the implementation implied abstracting the logic performed in the component `KeyboardAwareFlatList` to avoid redundancy and benefit from previous work. In this regard, the following hooks have been created to expose new scroll functions:
* [`useScrollToSection`](https://github.com/WordPress/gutenberg/pull/57273/commits/6ee8c01951cf181f1c67551bebaf44fcdef77cc9)
* [`useScrollToElement`](https://github.com/WordPress/gutenberg/pull/57273/commits/6d148b7fc91b265726042bdf3b53461fa8221b92)

Additionally, as part of the abstraction, the following changes have been applied:
* [Update logic to retrieve `KeyboardAwareFlatList` ref using `forwardRef`.](https://github.com/WordPress/gutenberg/pull/57273/commits/b6d7c69d8a6070f8e2991fd8aa2f3c5a590c92b5)
* [Replace hook `useScrollToTextInput` with `useScrollToSection`.](https://github.com/WordPress/gutenberg/pull/57273/commits/5f9e1c84859722502eb96890f217d5c113c0f65a)
* [Update Android's implementation of `KeyboardAwareFlatList` to expose the new scroll functions.](https://github.com/WordPress/gutenberg/pull/57273/commits/d9cbde9da33fe982c01c8c9e6fb0ade48f4defe4), which will be used in the auto-scroll functionality.
* [Unify the logic of both platforms in the hook `useScroll`](https://github.com/WordPress/gutenberg/pull/57273/commits/92750f293e49a80720e0108d12b93dac66d6fad3).

#### Retrieve inserted block's layout

I originally thought about using the layout data provided by [the block list context](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/block-list-context.native.js). But the fact, that at the moment of the insertion, it's not guaranteed that layout data is available. It made me lean toward using the native function [`measureLayout`](https://reactnative.dev/docs/0.71/direct-manipulation#measurelayoutrelativetonativecomponentref-onsuccess-onfail). This also will allow us to get the offset when inserting inner blocks, which it's currently not provided by the block list context.

## Testing Instructions

> [!NOTE]  
> It's recommended to use the following installable builds for testing:
> * [Android](https://github.com/wordpress-mobile/WordPress-Android/pull/19832#issuecomment-1865902351)
> * [iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/22266#issuecomment-1866014741)

> [!WARNING]  
> As mentioned in https://github.com/WordPress/gutenberg/pull/57273#issuecomment-1867856886, on Android the auto-scroll won't be triggered in all cases. We'll try to address this in the future.

**Preparation:**
1. Go to https://automattic.com/how-we-work/.
1. Copy the first part of the page (from How We Work until the end of the Our development process paragraph).
1. Open the app.
1. Create a new post.
1. Paste content.
1. Save the post.

### Case 1 - Insert a non-text-based block after opening a post
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Tap on the ➕ button.
1. Insert a non-text-based block like the Image block.
1. Expect the following:
    1. The content is scrolled to the end of the post.
    1. The block is added and fully visible on the screen.

### Case 2 - Insert a text-based block after opening a post
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Tap on the ➕ button.
1. Insert a text-based block like the Heading block.
1. Expect the following:
    1. The content is scrolled to the bottom of the post.
    1. The block is added and fully visible on the screen.
    1. The block is focused and the keyboard is open.
1. Type several words until a new line is created.
1. Expect that the content scrolls down to make the text visible.

### Case 3 - Insert a block to the beginning
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Scroll to the bottom of the content.
1. Select the last block.
1. Long press on the ➕ button.
1. Tap on _Add To Beginning_ option.
1. Insert any block.
1. Expect the following:
    1. The content is scrolled to the top of the post.
    1. The block is added and fully visible on the screen.
    1. If a text-based block is inserted, the block is focused and the keyboard is open.

### Case 4 - Insert a block after another located at the top screen
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Select the first block.
1. Tap on the ➕button.
1. Insert any block.
1. Expect the following:
    1. The scroll remains in the same position.
    1. The block is added and fully visible on the screen.
    1. If a text block is inserted, the block is focused and the keyboard is open.

### Case 5 - Insert a block after another located at the bottom screen
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Select the block located at the bottom of the screen (i.e. the visible area).
1. Tap on the ➕button.
1. Insert any block.
1. Expect the following:
    1. The content scrolled down.
    1. The block is added and fully visible on the screen.
    1. If a text block is inserted, the block is focused and the keyboard is open.

### Case 6 - Insert a block before another located at the top screen
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Scroll to the bottom of the content.
1. Select the block located at the top of the screen (i.e. the visible area).
1. Long press on the ➕button.
1. Tap on _Add Block Before_ option.
1. Insert any block.
1. Expect the following:
    1. The content scrolled up.
    1. The block is added and fully visible on the screen.
    1. If a text block is inserted, the block is focused and the keyboard is open.

### Case 7 - Insert a block before another located at the bottom screen
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Select the block located at the bottom of the screen (i.e. the visible area).
1. Long press on the ➕button.
1. Tap on _Add Block Before_ option.
1. Insert any block.
1. Expect the following:
    1. The scroll remains in the same position.
    1. The block is added and fully visible on the screen.
    1. If a text block is inserted, the block is focused and the keyboard is open.

### Case 8 - Paste content in the post title
1. Go to https://automattic.com/how-we-work/.
1. Copy the first part of the page (from How We Work until the end of the Our development process paragraph).
1. Open the app.
1. Create a new post.
1. Paste the content in the post title.
1. Expect the following:
    1. The scroll remains in the same position.
    1. All blocks have been added.

### Case 9 - Paste content in a text-based block
1. Go to https://automattic.com/how-we-work/.
1. Copy the first part of the page (from How We Work until the end of the Our development process paragraph).
1. Open the app.
1. Create a new post.
1. Select the first paragraph
1. Paste the content.
1. Expect the following:
    1. The content is scrolled to the end of the post.
    1. All blocks have been added.

### Case 10 - Copy block
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Select the first Image block.
1. Open the block options by tapping on the `...` button located on the toolbar.
1. Tap on Copy.
1. Long press on the ➕button.
1. Tap on _Add To End_ option.
1. Tap on _Copied block_.
1. Expect the following:
    1. The content is scrolled to the end of the post.
    1. The copied block is added and fully visible on the screen.

### Case 11 - Duplicate block
1. Open the app.
1. Open the post created after following the Preparation steps.
1. Select the first Image block.
1. Open the block options by tapping on the `...` button located on the toolbar.
1. Tap on Duplicate.
1. Expect the following:
    1. The scroll remains in the same position.
    1. The block is duplicated.

### Case 12 - Landscape mode
1. Rotate the device to landscape mode.
1. Repeat all the above cases.
1. Rotate between portrait and landscape modes.
1. Expect the scroll to remain in the same position.

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 

| Android | iOS |
|--------|--------|
| <video src=https://github.com/WordPress/gutenberg/assets/14905380/a1599176-2410-422e-8e4e-21bf7738debf> | <video src=https://github.com/WordPress/gutenberg/assets/14905380/f4e60b9e-746c-4449-87b2-2015af6b0ff1> |

### CODE REVIEW

**Comment by geriux** [comment]
Hey @fluiddot 👋 

I'm going through the test cases but I have a question for Case 8:

> Case 8 - Paste content
> Follow the Preparation steps.
> After pasting the content, expect the following:
> The scroll remains in the same position.
> All blocks have been added.

What do you mean by "The scroll remains in the same position."? Because I think that will depend if you paste the content from the title or if you paste it from the first Paragraph block. If you paste it from the Title, the scroll remains the same but if you paste it in the Paragraph block it scrolls down. This behavior was already like that I just want to make sure which steps should I take for this test case specifically. Thank you!

**Comment by fluiddot** [comment]
> What do you mean by "The scroll remains in the same position."? Because I think that will depend if you paste the content from the title or if you paste it from the first Paragraph block. If you paste it from the Title, the scroll remains the same but if you paste it in the Paragraph block it scrolls down. This behavior was already like that I just want to make sure which steps should I take for this test case specifically. Thank you!

Good observation. I'll split the test case into two:
- Paste content in the post title.
- Paste content in a text-based block.

Thanks for the call out 🙇 !

**Comment by fluiddot** [comment]
> > What do you mean by "The scroll remains in the same position."? Because I think that will depend if you paste the content from the title or if you paste it from the first Paragraph block. If you paste it from the Title, the scroll remains the same but if you paste it in the Paragraph block it scrolls down. This behavior was already like that I just want to make sure which steps should I take for this test case specifically. Thank you!
> 
> Good observation. I'll split the test case into two:
> 
> * Paste content in the post title.
> * Paste content in a text-based block.
> 
> Thanks for the call out 🙇 !

@geriux I've just updated the testing instructions to reflect the two mentioned test cases.

**Comment by geriux** [comment]
Sharing my initial results so far:

- All test cases on iOS worked perfectly ✅ 
- On Android, all tests worked as expected in portrait mode ✅ 
- ⚠️ On Android using the device in landscape mode I'm having issues with the first test case:

> Case 1 - Insert a non-text-based block after opening a post

https://github.com/WordPress/gutenberg/assets/4885740/f52a8813-dde4-4fbc-8a6b-24a072637692

Steps to reproduce:

- Open the post created after following the Preparation steps.
- Add a Spacer block
- The list doesn't scroll to the newly added block

Are you able to reproduce? I even waited a bit for the content to be fully rendered but no luck so far. On Portrait mode, it does work as expected 🎉 

I also checked with a text-based block and it doesn't auto-scroll, but It looks like it's an issue unrelated to these changes because I can also reproduce on the production app. I was expecting that at least for text-based blocks it would auto-scroll.

File: packages/components/src/mobile/keyboard-aware-flat-list/index.android.js [hunk 1]
```diff
@@ -2,26 +2,69 @@
...
+	const getFlatListRef = useCallback(
+		( flatListRef ) => {
+			// On Android, we get the ref of the associated scroll
+			// view to the FlatList.
+			scrollViewRef.current = flatListRef?.getNativeScrollRef();
+		},
+		[ scrollViewRef ]
+	);
+
+	useImperativeHandle( ref, () => {
```
Comment (geriux) [inline packages/components/src/mobile/keyboard-aware-flat-list/index.android.js:51]: Love this! 😍

**Comment by fluiddot** [comment]
Thanks @geriux for reviewing the PR 🙇 !

> Are you able to reproduce? I even waited a bit for the content to be fully rendered but no luck so far. On Portrait mode, it does work as expected 🎉

That's interesting, in previous tests I've made it seems to work nice. However, I've re-tested using a longer post (as the one proposed in the Preparation steps) and I encountered the issue. I'm afraid this might be related to the list virtualization, which AFAIK is only enabled on Android, because if you scroll down a bit the auto-scroll is triggered (see attached video).

I'll check further but this might be a case to be handled differently. I'm thinking to use `scrollToEnd` as we know that the block will be inserted at the end of the content 🤔.

https://github.com/WordPress/gutenberg/assets/14905380/8dccb85b-626d-4daa-a1ba-7d4104d09b13

> I also checked with a text-based block and it doesn't auto-scroll, but It looks like it's an issue unrelated to these changes because I can also reproduce on the production app. I was expecting that at least for text-based blocks it would auto-scroll.

I presume the problem I shared above affects all blocks, including text-based blocks.

**Comment by fluiddot** [comment]
Following [this comment](https://github.com/WordPress/gutenberg/pull/57273#issuecomment-1867574012) and as discussed internally, due to the list virtualization on Android, the auto-scroll won't be triggered in all cases.

In https://github.com/WordPress/gutenberg/pull/57273/commits/87a7a12c7a0b9cf2c7839417be38b64664947b4d, I addressed an issue on Android related to triggering the auto-scroll multiple times for the same block, which was caused by the list virtualization.

@geriux let me know if you could take a look at the recent changes, thanks 🙇 !

**Comment by geriux** [comment]
Leaving a note here to mention that I encountered this issue https://github.com/WordPress/gutenberg/issues/57355 that will be addressed in https://github.com/WordPress/gutenberg/pull/57279 while testing on Android but it does not block this PR.

**Comment by geriux** [comment]
Leaving a note here to mention that I encountered this issue https://github.com/WordPress/gutenberg/issues/57355 that will be addressed in https://github.com/WordPress/gutenberg/pull/57279 while testing on Android but it does not block this PR.

**Review by geriux — APPROVED** [review]
LGTM! Great work!! 🤩 

I've followed the test cases for both platforms using the latest builds and it worked as expected taking into account the limitations on Android.

The code changes look go to me as well, thank you for unifying the logic to re-use some of the code we had 🥇

**Pull request merged**